### PR TITLE
Incremental PR for STAC Point Generation

### PR DESCRIPTION
**Pull request by harryle95**
Basic Features

- [x] Specifying X and Y coords and Projection for the csv point raw data 
- [x] Specifying the band information (csv fields) for efficient loading 
- [x] Specifying the csv fields for partitioning the points to item groups (if no field is provided, the whole collection will have exactly one point)
- [x] Extract spatial and temporal extent for STAC items based on csv field values
- [x] Set asset links for item and collection 

Things not covered 
- [ ] Testing
- [ ] Conforming to the previous interface

### CODE REVIEW

File: stac_generator/typing.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+DateTimeT = datetime.datetime
+BBoxT = list[NumberT]  # Either 2D or 3D BBoxType
+TimeExtentT = tuple[DateTimeT | None, DateTimeT | None]  # (start_datetime, end_datetime) format
+CSVMediaType = "text/csv"  # https://www.rfc-editor.org/rfc/rfc7111
+ExcelMediaType = "application/vnd.ms-excel"  # https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+FrameT = pd.DataFrame
+
+# GeoJSON types
+
+GeometryT = Literal[
```
Comment (JosephUoA) [inline stac_generator/typing.py]: Is it possible to use an existing definition for geometry types, such as [shapely geometries](https://shapely.readthedocs.io/en/stable/geometry.html)?

File: stac_generator/point_stac_generator.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    return "".join(parts)
+
+
+def get_path(path: str | Path) -> Path:
+    if isinstance(path, Path):
+        return path
+    return Path(path)
+
+
+def calculate_spatial_extent(
```
Comment (JosephUoA) [inline stac_generator/point_stac_generator.py]: Like you said this can probably be solved with geopandas. I think its a good idea to use geopandas instead of regular pandas whenever there is a spatial component to the data. 

```
import geopandas as gpd

gdf = gpd.read_file(filepath)
print(gdf.geometry.total_bounds())

> array([min_X, min_Y, max_X, max_Y])
```

File: stac_generator/point_stac_generator.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    Y_coord: str,
+    T_coord: str | None = None,
+    date_format: str = "ISO8601",
+    bands: Sequence[str] | Sequence[PointBandInfo] | None = None,
+    item_group: Sequence[str] | None = None,
+) -> FrameT:
+    parse_dates = [T_coord] if isinstance(T_coord, str) else False
+    usecols = None
+    # If band info is provided, only read in the required columns + the X and Y coordinates
+    if bands:
```
Comment (JosephUoA) [inline stac_generator/point_stac_generator.py]: My gut feeling is to refer to this variable `bands` with a different name to avoid confusing it with raster data. Maybe `data_variables`? @dhobern do you have any thoughts? I can explain it better over a call if necessary.

File: soilPointDataAllSites_4ML.csv
Comment (JosephUoA) [inline soilPointDataAllSites_4ML.csv:1]: We probably should not commit this file.

File: stac_generator/point_stac_generator.py (hunk 3 above)
Comment (dhobern) [reply stac_generator/point_stac_generator.py]: Personally, I think of the data_variables as bands when we start adding them as layers onto a map. They are just as much "bands" as NDVI is. But I don't have a strong opinion, particularly if it is just a local name for this method as it seems to be here.

Comment (dhobern) [reply stac_generator/point_stac_generator.py]: I see bands is a parameter to the constructor, so not just the local name here, but I remain easy either way.

File: stac_generator/point_stac_generator.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    FrameT,
+    GeometryObj,
+    PointBandInfo,
+    TimeExtentT,
+)
+
+
+# Helper functions
+def ensure_name_suffix(name: str, suffix: str) -> str:
+    """Ensure that name ends with suffix.
```
Comment (JosephUoA) [inline stac_generator/point_stac_generator.py]: Sorry to be an absolute pain but could we use reST style docstrings in MCCN repos? Does this conflict with mkdocs?

```
This is a reST style.

:param param1: this is a first param
:param param2: this is a second param
:returns: this is a description of what is returned
:raises keyError: raises an exception
```

Comment (harryle95) [reply stac_generator/point_stac_generator.py]: Shouldn't affect mkdocs, but I will need to find a way to configure vscode docstring extension. I will give it a try

File: stac_generator/typing.py (hunk 1 above)
Comment (harryle95) [reply stac_generator/typing.py]: Addressed in the latest commit

File: stac_generator/point_stac_generator.py (hunk 2 above)
Comment (harryle95) [reply stac_generator/point_stac_generator.py]: Resolved in the latest commit

File: soilPointDataAllSites_4ML.csv
Comment (harryle95) [reply soilPointDataAllSites_4ML.csv:1]: File removed, but add a subset of this file as examples under `example/csv`

File: stac_generator/sensor_stac_generator.py
Comment (JosephUoA) [inline stac_generator/sensor_stac_generator.py:1]: Lets just delete this file

File: tests/test_generator.py [hunk 5]
```diff
@@ -11,16 +11,16 @@ def test_generator() -> None:
     # data_type = data_file.stem.split("_")[0]
     # assert data_type == "drone"
     # location_file = Path("tests/test_data/drone_test_files.csv")
-    # # Create the STAC catalog.
-    # os.environ["STAC_API_URL"] = "placeholder"
+    # # Create the Stac catalog.
+    # os.environ["Stac_API_URL"] = "placeholder"
```
Comment (JosephUoA) [inline tests/test_generator.py]: Lets not allow the linter or whatever changed this to decapitalize then env variable names. The comments should not have changed either but its less important.

File: stac_generator/types.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+"""tuple of start and end dates"""
+TimeExtentT = tuple[DateTimeT | None, DateTimeT | None]  # (start_datetime, end_datetime) format
+
+"""<a href="https://www.rfc-editor.org/rfc/rfc7111">CSV MIME type</a>"""
+CSVMediaType = "text/csv"
+"""<a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types">Excel MIME type</a>"""
+ExcelMediaType = "application/vnd.ms-excel"
+
+"""Pandas DataFrame"""
+PDFrameT = pd.DataFrame
```
Comment (JosephUoA) [inline stac_generator/types.py]: Some of these types I can get behind but some I dont understand why you would obfuscate the code. Why did you rename the types for the DataFrames in particular? Even the NumberT and DateTimeT I dont really like.

File: stac_generator/generator.py [hunk 7]
```diff
@@ -1,31 +1,34 @@
...
-            raise ValueError("Environment variable STAC_API_URL must be defined.")
+            raise ValueError("Environment variable Stac_API_URL must be defined.")
         self.data_type = data_type
         self.data_file = data_file
         self.location_file = location_file
         self.standard_file = f"./standards/{self.data_type}_standard.csv"
-        self.items: List[pystac.Item] = []
-        self.catalog: Optional[pystac.Catalog] = None
-        self.collection: Optional[pystac.Collection] = None
+        self.items: list[pystac.Item] = []
```
Comment (JosephUoA) [inline stac_generator/generator.py]: I guess I need to remember that the typing module is deprecated, for some reason this is hard for me to get used to.

File: stac_generator/generator.py [hunk 8]
```diff
@@ -34,32 +37,36 @@ def validate_data(self) -> bool:
...
+        self.write_collection_to_api()
+        self.write_items_to_api()
 
-    @abstractmethod
     def validate_stac(self) -> bool:
-        """Check that the generated STAC is valid."""
-        raise NotImplementedError
+        if self.catalog and not self.catalog.validate():
+            return False
+        return not (self.collection and not self.collection.validate())
```
Comment (JosephUoA) [inline stac_generator/generator.py]: I would prefer not to refactor the boolean logic like this. While not particularly difficult to parse, it is more difficult than what was there before.

File: stac_generator/generator_factory.py [hunk 9]
```diff
@@ -1,15 +1,16 @@
 from stac_generator.drone_stac_generator import DroneStacGenerator
-from stac_generator.sensor_stac_generator import SensorStacGenerator
 from stac_generator.generator import StacGenerator
+from stac_generator.sensor_stac_generator import SensorStacGenerator
+
+__all__ = ("StacGeneratorFactory",)
```
Comment (JosephUoA) [inline stac_generator/generator_factory.py]: Why did you add this?

File: stac_generator/csv/utils.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+    Returns a `shapely.Point` or `shapely.MultiPoint` depending on the number
+    of unique points in the dataframe.
+
+    :param df: source dataframe
+    :type df: FrameT
+    :return: shapely geometry object
+    :rtype: Point | MultiPoint
+    """
+    points: Sequence[Point] = df["geometry"].unique()
```
Comment (JosephUoA) [inline stac_generator/csv/utils.py:141]: geoDataFrame can have polygons or other Geometry types in the geometry column. We can probably assume that we are taking only points when reading from csv but maybe we can explicit that this only works for this case.

File: stac_generator/base/__init__.py
Comment (JosephUoA) [inline stac_generator/base/__init__.py:1]: I gotta be honest, I have no idea whats going on with this and the next 9 files. Probably have to explain it to me.

File: tests/test_generator.py (hunk 5 above)
Comment (harryle95) [reply tests/test_generator.py]: A lot of the refactoring was done cause I wanted to check if the CI works well. I will reconfigure the linter to not raise an error for this.

File: stac_generator/types.py (hunk 6 above)
Comment (harryle95) [reply stac_generator/types.py]: It's laziness mainly - i.e. PDFrameT is 2 strokes (P + Tab) whereas pd.DataFrame is 4 strokes (p + d + . + D). Same for NumberT and DateTimeT. I can revert the types no problem.

File: stac_generator/generator.py (hunk 8 above)
Comment (harryle95) [reply stac_generator/generator.py]: Same as [here](https://github.com/aus-plant-phenomics-network/stac-generator/pull/7#discussion_r1792835754)

File: stac_generator/generator_factory.py (hunk 9 above)
Comment (harryle95) [reply stac_generator/generator_factory.py]: It's from one of the precommit hooks. It ensure that all files have __all__ i.e. choose what name to export

File: stac_generator/types.py (hunk 6 above)
Comment (JosephUoA) [reply stac_generator/types.py]: Ok please revert then. I think a few more keypresses are worth the legibility!

File: stac_generator/generator.py (hunk 8 above)
Comment (JosephUoA) [reply stac_generator/generator.py]: The linter did this?! What a jerk

File: stac_generator/generator_factory.py (hunk 9 above)
Comment (JosephUoA) [reply stac_generator/generator_factory.py]: Never seen a hook like that before. By default you export everything without a leading underscore so I don't see why its necessary.

File: stac_generator/base/__init__.py
Comment (harryle95) [reply stac_generator/base/__init__.py:1]: SourceConfig is the schema of the csv config that contains the source location. An example of this file is `example/csv/source_config.csv`. SourceConfig contains information for generating the source asset - i.e. its href, endpoint, etc, and other processing information for processing the source file and append to the `Item.properties` attribute. For example, for processing the csv data, I would need to know the X, Y columns etc. This info should be found in the source config. Any file extension config should inherit this class. 

`CatalogConfig` and `CollectionConfig` are metadata at collection and catalog level - i.e. license, keyword, local href for serialisation and api_endpoint for pushing to STAC api endpoint. 

IODriver is the class that handles reading in and processing the SourceConfig. For instance, reading a csv and a geotiff should be handled differently and this class should be subclassed to handle the discrepancy. It could be merged with StacCatalog class but TBD.

utils files contain utility methods for each file extension. The global utils have utility functions like extracting the collection bbox from the bboxes of all its child item, and extracting the time extent in a similar manner. The csv utils have utils for reading csv file, converting to geopandas so on and so forth. 

generator contains the STAC generator class that handles generating items/collection/catalog, saving catalog as json file to disk, and POST to a STAC API endpoint - i.e. basically the previous StacGenerator class but with the added config params. There is only one abstractmethod now which is `create_item_from_config`. Handlers for different file extension now only need to define this method.

File: docs/driver.md
Comment (JosephUoA) [inline docs/driver.md:1]: Had a long think about it and I've come to the conclusion that we should not commit the generated markdown files. Instead add instructions for how to generate the docs in the README.

In fact the instructions to do many things probably need to be added to the README at this point:
1. How to generate docs
2. How to run CI locally
3. How to use the tool (this ones for me to do)

Theres probably more.

File: stac_generator/generator_factory.py (hunk 9 above)
Comment (harryle95) [reply stac_generator/generator_factory.py]: It's for type checking tools like mypy. It's more about explicitly forbidding private variables. Without __all__, when you import a private method, mypy would not say anything. If you don't put the private method in __all__ and try to import it, mypy will raise an issue. It's part of precommit hook, and it's just calling `make lint`. I can disable this if you want.

File: pdm.lock
Comment (JosephUoA) [inline pdm.lock:1]: You can't just go and do something like change the package manager on a whim. I understand that something was broken with Poetry but this type of change has a much larger flow on effect to all the developers working on this and other repos. I can't even really tell what other changes in this PR are motivated by the switch in package manager. Clearly the makefile somewhat, maybe the github actions split up. Not sure what to recommend from here.

File: docs/driver.md
Comment (harryle95) [reply docs/driver.md:1]: The md files are not the generated files. It's the template that specifies how mkdocs should generate the docs. The generated files are not commited. Will commit the readme now.

File: stac_generator/base/generator.py [hunk 11]
```diff
@@ -2,57 +2,76 @@
...
 from stac_generator.base.utils import extract_spatial_extent, extract_temporal_extent
-from stac_generator.types import StacEntityT
-
-__all__ = (
-    "StacGenerator",
-    "StacLoader",
-)
 
 
 class StacGenerator(Generic[T]):
```
Comment (JosephUoA) [inline stac_generator/base/generator.py]: Still no ABC?

Comment (harryle95) [reply stac_generator/base/generator.py]: One ABC coming right up

**Review by JosephUoA — APPROVED** [review]

**Comment by JosephUoA** [comment]
@harryle95 Can you please resolve conflicts and merge when ready

File: .vscode/settings.json
Comment (JosephUoA) [inline .vscode/settings.json:1]: Do you want to .gitignore this?

File: pyproject.toml [hunk 12]
```diff
@@ -119,6 +92,7 @@ lint.ignore = [
     "ISC001",  # Ruff formatter incompatible
     "A002",    # Argument id shadowing scope id
     "ERA001",  # Commented out code
+    "G004",    # flake8-logging-format - logging-f-string
```
Comment (JosephUoA) [inline pyproject.toml:95]: I think we should not log f-strings 😂 . Something an old colleague drilled into me.

File: tests/base/test_utils.py
Comment (JosephUoA) [inline tests/base/test_utils.py:1]: Good job with these tests
